src/main/scala/async/Async.scala

```diff
@@ -180,78 +178,44 @@ object Async:
           it.next.poll(listener)
         found
 
-      // def poll(k: Listener[T]): Boolean =
-      //   val it = sources.iterator
-      //   var found = false
-
-      //   val listener = new Listener[T]:
-      //     def tryLock() =
-      //       val result = k.tryLock()
-      //       // If tryLock is called, the source has - to the best of its
-      //       // knowledge - an item available. But the upstream listener k
-      //       // refuses to take any -> we assume there would have been one.
-      //       if result == Listener.Gone then found = true
-      //       result
-
-      //     def release(until: Listener.ReleaseBoundary) = k
-
-      //     def complete(data: T) =
-      //       k.complete(data)
-      //       found = true
-      //   end listener
-
-      //   while it.hasNext && !found do
-      //     it.next.poll(listener)
-      //   found
-
       def onComplete(k: Listener[T]): Unit =
-        val listener = new Listener.ForwardingListener[T](this, k) with Listener.LockingListener { self =>
-          var foundBefore = false
-
-          val listenerPartial = new Listener.NumberedSublock {
-            def number: Long = self.number
-            def tryLock() =
-              self.lock()
-              val result =
-                if foundBefore then
-                  self.unlock()
-                  Listener.Gone
-                else
-                  val result = k.tryLock()
-                  if result == Listener.Gone then self.unlock()
-                  result
-              end result
-              // as soon as we once return None, we are completed -> drop everywhere
-              if result == Listener.Gone then sources.foreach(_.dropListener(self))
-              result
-          } // end listenerPartial: NumberedSublock
-
-          def tryLock() =
-            if foundBefore then Listener.Gone
-            else listenerPartial
-
-          def release(until: Listener.ReleaseBoundary) =
-            if until != listenerPartial then
-              self.unlock()
-              k
-            else null
-
-          def complete(data: T) =
-            k.completeNow(data)
-            foundBefore = true
-            self.unlock()
-            sources.foreach(_.dropListener(self))
+        val listener = new Listener.ForwardingListener[T](this, k) with NumberedLock with Listener.ListenerLock with Listener.PartialLock { self =>
```

Comment:
The wrapped race-listener currently does not drop itself if the downstream listener fails (see `ListenerBehavior.race failed without wait`).